Support fetching and evaluating hospitalization forecasts

nmdefries opened the pull request:
### Description
Support fetching and evaluating hospitalization forecasts.

### Changes
- Update `get_covidhub_predictions` documentation and default signals list
- Update documentation, default signals list, and logic defining `signal`, `response`, and `data_source` for a prediction for `get_zoltar_predictions`, `get_covidhub_predictions_alt`, and `get_forecaster_predictions`
- Remove trailing whitespace
- Switch filter in `get_zoltar_predictions` to remove all "drop" responses instead of all "hosp" responses.

dajmcdon commented on R-packages/evalcast/R/get_covidhub_predictions.R:
```diff
@@ -243,31 +245,38 @@ get_forecaster_predictions <- function(covidhub_forecaster_name,
              incidence_period = if_else(
                .data$incidence_period == "wk", "epiweek","day"),
              inc = if_else(.data$inc == "inc", "incidence", "cumulative"),
-             response = case_when(.data$response == "death" ~ "deaths", 
+             response = case_when(.data$response == "death" ~ "deaths",
                                   .data$response == "case" ~ "confirmed",
-                                  TRUE ~ "drop",),
-             signal = paste(.data$response, .data$inc, "num", sep="_"),
-             data_source = if_else(.data$response=="drop", "drop", "jhu-csse"),
+                                  .data$response == "hosp" ~ "hosp",
+                                  TRUE ~ "drop"),
+             data_source = case_when(.data$response == "deaths" ~ "jhu-csse",
+                                  .data$response == "confirmed" ~ "jhu-csse",
+                                  .data$response == "hosp" ~ "hhs",
+                                  TRUE ~ "drop"),
+             signal = case_when(
+               .data$data_source == "jhu-csse" ~ paste(.data$response, .data$inc, "num", sep="_"),
+               .data$data_source == "hhs" & .data$inc == "incidence" ~ "confirmed_admissions_covid_1d",
```
Can you extract this chunk (and the similar one marked below) into a separate (testable) function?

dajmcdon commented on R-packages/evalcast/R/get_covidhub_predictions.R:
```diff
@@ -373,11 +383,18 @@ get_forecaster_predictions_alt <- function(covidhub_forecaster_name,
+           response = case_when(.data$response == "death" ~ "deaths",
                                 .data$response == "case" ~ "confirmed",
-                                TRUE ~ "drop",),
-           signal = paste(.data$response, .data$inc, "num", sep="_"),
-           data_source = if_else(.data$response=="drop", "drop", "jhu-csse"),
+                                .data$response == "hosp" ~ "hosp",
+                                TRUE ~ "drop"),
+           data_source = case_when(.data$response == "deaths" ~ "jhu-csse",
+                                   .data$response == "confirmed" ~ "jhu-csse",
+                                   .data$response == "hosp" ~ "hhs",
+                                   TRUE ~ "drop"),
+           signal = case_when(
+             .data$data_source == "jhu-csse" ~ paste(.data$response, .data$inc, "num", sep="_"),
+             .data$data_source == "hhs" & .data$inc == "incidence" ~ "confirmed_admissions_covid_1d",
+             TRUE ~ "drop"),
```
See above comment.

dajmcdon commented on R-packages/evalcast/R/get_covidhub_predictions.R:
```diff
@@ -391,14 +408,14 @@ get_forecaster_predictions_alt <- function(covidhub_forecaster_name,
              value = as.double(value)) %>%
       relocate(.data$ahead, .data$location, .data$quantile, .data$value,
                .data$forecaster, .data$forecast_date, .data$data_source,
-               .data$signal, .data$target_end_date, 
+               .data$signal, .data$target_end_date,
                .data$incidence_period) %>%
       filter(.data$response != "drop", .data$type %in% forecast_type,
              .data$incidence_period %in% incidence_period,
              .data$signal %in% signal) %>%
       select(.data$ahead, .data$location, .data$quantile, .data$value,
                .data$forecaster, .data$forecast_date, .data$data_source,
-               .data$signal, .data$target_end_date, 
+               .data$signal, .data$target_end_date,
                .data$incidence_period)
```
Maybe this part can go in too??

dajmcdon commented on R-packages/evalcast/R/get_zoltar_predictions.R:
```diff
@@ -117,19 +120,26 @@ get_zoltar_predictions <- function(forecaster_names = NULL,
+      response = case_when(.data$response == "death" ~ "deaths",
                            .data$response == "case" ~ "confirmed",
-                           TRUE ~ "drop",),
-      signal = paste(.data$response, .data$inc, "num", sep="_"),
-      data_source = if_else(.data$response == "drop", "drop", "jhu-csse"),
+                           .data$response == "hosp" ~ "hosp",
+                           TRUE ~ "drop"),
+      data_source = case_when(.data$response == "deaths" ~ "jhu-csse",
+                              .data$response == "confirmed" ~ "jhu-csse",
+                              .data$response == "hosp" ~ "hhs",
+                              TRUE ~ "drop"),
+      signal = case_when(
+        .data$data_source == "jhu-csse" ~ paste(.data$response, .data$inc, "num", sep="_"),
+        .data$data_source == "hhs" & .data$inc == "incidence" ~ "confirmed_admissions_covid_1d",
+        TRUE ~ "drop"),
```
Same thing again.

dajmcdon reviewed (changes requested):
I marked 3 locations where we're reusing largely similar code. Can we extract that all into a single, testable, function (along with tests)? It looks like yes, but I haven't stared at it for long enough recently enough.

Otherwise, everything looks great.

nmdefries commented:
Added a new set of predictions processing functions to deduplicate `get_zoltar_predictions`, `get_covidhub_predictions`, and `get_covidhub_predictions_alt`. The new functions:

* Separate and parse the `target` field, and create derived columns (`data_source`, etc).
* Remove invalid/irrelevant predictions.
* Keep only predication card columns in desired order. Note: I removed a `relocate` command in `get_covidhub_predictions_alt`  that didn't appear to be doing anything (output for a toy example didn't change with/without) since a following `select` reorders the columns anyway.
* Create the `geo_value` column from `location` and in the correct order.

dajmcdon commented on R-packages/evalcast/R/predictions_processing_utils.R:
```diff
@@ -0,0 +1,93 @@
+#'   "confirmed_admissions_covid_1d".
+#'
+#' @return predictions dataframe with irrelevant rows removed
+filter_predictions <- function(predictions,
+                               forecast_type,
+                               incidence_period,
+                               signal) {
+  # Filter does not use variables in global environment when variable names is
+  # the same as a column name, so do a workaround.
+  forecast_type <- enquo(forecast_type)
+  incidence_period <- enquo(incidence_period)
+  signal <- enquo(signal)
+  
+  predictions %>%
+    filter(.data$response != "drop",
+           .data$type %in% (!!forecast_type),
+           .data$incidence_period %in% (!!incidence_period),
+           .data$signal %in% (!!signal))
+}
```
I can't quite tell since you're passing arguments through a few functions. Can you use the `{{ forecast_type }}` syntax rather than the `enquo()` workaround? See `?rlang::nse-force`.

nmdefries replied:
It worked! Thanks for the cleaner suggestion.

dajmcdon reviewed (approved):
Thanks!! This all looks great. I left one comment for a possible simplification of your workaround, but I'm not sure if my suggestion will actually work given the nesting of functions. Approved otherwise.
